modules/apps/29-fee/ibc_module.go

```diff
@@ -0,0 +1,168 @@
+	ibcexported "github.com/cosmos/ibc-go/modules/core/exported"
+)
+
+// IBCModule implements the ICS26 callbacks for the fee middleware given the fee keeper and the underlying application.
+type IBCModule struct {
+	keeper keeper.Keeper
+	app    porttypes.IBCModule
+}
+
+// NewIBCModule creates a new IBCModule given the keeper and underlying application
+func NewIBCModule(k keeper.Keeper, app porttypes.IBCModule) IBCModule {
+	return IBCModule{
+		keeper: k,
+		app:    app,
+	}
+}
+
+// OnChanOpenInit implements the IBCModule interface
+func (im IBCModule) OnChanOpenInit(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version string,
+) error {
+	feeVersion, appVersion := channeltypes.SplitChannelVersion(version)
+	if feeVersion != "" {
+		if feeVersion != types.Version {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected: %s, got: %s", types.Version, feeVersion)
+		}
+
+		im.keeper.SetFeeEnabled(ctx, portID, channelID)
+	}
+
+	// call underlying app's OnChanOpenInit callback with the appVersion
+	return im.app.OnChanOpenInit(ctx, order, connectionHops, portID, channelID,
+		chanCap, counterparty, appVersion)
+}
+
+// OnChanOpenTry implements the IBCModule interface
+func (im IBCModule) OnChanOpenTry(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version,
+	counterpartyVersion string,
+) error {
+	feeVersion, appVersion := channeltypes.SplitChannelVersion(version)
+	cpFeeVersion, cpAppVersion := channeltypes.SplitChannelVersion(counterpartyVersion)
+
+	if feeVersion != "" || cpFeeVersion != "" {
+		if feeVersion != types.Version {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected: %s, got: %s", types.Version, feeVersion)
+		}
+		if cpFeeVersion != feeVersion {
+			return sdkerrors.Wrapf(types.ErrInvalidVersion, "expected counterparty version: %s, got: %s", types.Version, cpFeeVersion)
+		}
+
+		im.keeper.SetFeeEnabled(ctx, portID, channelID)
+	} else if im.keeper.IsFeeEnabled(ctx, portID, channelID) {
+		// return error if a previous ChanInit set fee enabled but subsequent OpenTry does not have fee enabled
+		return sdkerrors.Wrapf(types.ErrInvalidVersion, "previous INIT call (crossing hellos) had fee version set but OpenTry call does not set fee version: %s", version)
+	}
+	// call underlying app's OnChanOpenTry callback with the app versions
+	return im.app.OnChanOpenTry(ctx, order, connectionHops, portID, channelID,
+		chanCap, counterparty, appVersion, cpAppVersion)
+}
+
+// OnChanOpenAck implements the IBCModule interface
+func (im IBCModule) OnChanOpenAck(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	counterpartyVersion string,
+) error {
+	cpAppVersion := counterpartyVersion
+	if im.keeper.IsFeeEnabled(ctx, portID, channelID) {
```

Comment:
No because in that case the entire counterparty version `fee29-1:ics20-1` would be passed down to transfer which would error.

Though it's a good thing to test. I added a test case to prove it.

Reply:
We don't want future relayers to override the initial decisions of past relayers.

Future relayers should only complete the handshake as initial relayers intended it

Reply:
That makes sense. Can you add a inline comment to the code?